### PR TITLE
[Feature] Refactor Estimator for computing FLOPs/Params/Latency.

### PR DESCRIPTION
**Pull request by gaoyang07**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Refactor Estimator for computing FLOPs/Params/Latency.

## Modification

1. Add `ResourceEstimator` to estimate model resources.
2. Refactor `mmcv.flops_counter` as `flops_params_counter`.
3. Add `latency_counter`.
4. Add counters for common op counters, e.g. `ConvCounter`.
5. Add `EstimateResourcesHook`.
6. Add UT for `flops_params_counter` & `ResourceEstimator`.
7. Remove old `FlopsEstimator` in mmrazor.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/registry/registry.py [hunk 1]
```diff
@@ -111,3 +111,6 @@ def build_razor_model_from_cfg(
 VISUALIZERS = Registry('visualizer', parent=MMENGINE_VISUALIZERS)
 # manage visualizer backend
 VISBACKENDS = Registry('vis_backend', parent=MMENGINE_VISBACKENDS)
+
+ESTIMATOR = Registry('estimator')
```
Comment (sunnyxiaohu) [inline mmrazor/registry/registry.py:115]: ESTIMATOR -> ESTIMATORS

File: mmrazor/structures/estimator/latency.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                    print(
+                        f'Done image [{i + 1:<3}/ {max_iter}], '
+                        f'fps: {fps:.1f} img / s, '
+                        f'times per image: {1000 / fps:.1f} ms / img',
+                        flush=True)
+
+        if (i + 1) == max_iter:
+            fps = (i + 1 - num_warmup) / pure_inf_time
+            if PRINT:
+                print(
```
Comment (sunnyxiaohu) [inline mmrazor/structures/estimator/latency.py:83]: use logger to print, with debug logger level.

File: mmrazor/structures/estimator/latency.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                                    log_interval))
+
+    if repeat_num > 1:
+        fps_list_ = [round(fps, 1) for fps in fps_list]
+        times_pre_image_list_ = [round(1000 / fps, 1) for fps in fps_list]
+        mean_fps_ = sum(fps_list_) / len(fps_list_)
+        mean_times_pre_image_ = sum(times_pre_image_list_) / len(
+            times_pre_image_list_)
+        if PRINT:
+            print(
```
Comment (sunnyxiaohu) [inline mmrazor/structures/estimator/latency.py:31]: use logger to print, with debug logger level.

File: mmrazor/structures/estimator/base_estimator.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+from mmrazor.registry import ESTIMATOR
+from .flops_params_counter import (get_model_complexity_info,
+                                   params_units_convert)
+from .latency import repeat_measure_inference_speed
+from .op_spec_counters import BaseCounter
+
+
+@ESTIMATOR.register_module()
+class BaseEstimator(metaclass=ABCMeta):
+    """Evaluator for calculating the accuracy and resources consume. Accuracy
```
Comment (sunnyxiaohu) [inline mmrazor/structures/estimator/base_estimator.py:17]: update docstring, including necessary ``Notes``.

File: mmrazor/structures/estimator/base_estimator.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                 default_shape: Tuple = (1, 3, 224, 224),
+                 units: str = 'M',
+                 disabled_counters: BaseCounter = None):
+        assert len(default_shape) in [3, 4, 5], \
+            f'Unsupported shape: {default_shape}'
+        self.default_shape = default_shape
+        self.units = units
+        self.disabled_counters = disabled_counters
+
+    def evaluate(
```
Comment (sunnyxiaohu) [inline mmrazor/structures/estimator/base_estimator.py:39]: `evaluate` -> `estimate`

File: mmrazor/registry/registry.py (hunk 1 above)
Comment (gaoyang07) [reply mmrazor/registry/registry.py]: Done.

File: mmrazor/structures/estimator/latency.py (hunk 2 above)
Comment (gaoyang07) [reply mmrazor/structures/estimator/latency.py]: Done.

File: mmrazor/structures/estimator/latency.py (hunk 3 above)
Comment (gaoyang07) [reply mmrazor/structures/estimator/latency.py]: Done.

File: mmrazor/structures/estimator/base_estimator.py (hunk 5 above)
Comment (gaoyang07) [reply mmrazor/structures/estimator/base_estimator.py]: Done.

File: mmrazor/structures/estimator/op_spec_counters/base_counter.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from abc import ABCMeta, abstractclassmethod
+
+
+class BaseCounter(object, metaclass=ABCMeta):
```
Comment (sunnyxiaohu) [inline mmrazor/structures/estimator/op_spec_counters/base_counter.py:5]: Point that `XXModuleCounter` is responsible for `XXModule`, which could refers to flops_params_counter::`get_counter_type()`.

File: mmrazor/engine/runner/evaluator_val_loop.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        # TODO confirm the state judgement.
+        if hasattr(model, 'is_supernet') and model.is_supernet:
+            model = self.export_subnet(model)
+
+        resource_results = self.estimator.estimate(
+            model=model, resource_args=resource_args)
+
+        return resource_results
+
+    def export_subnet(self, model):
```
[the remark this answers is not in the document]
Comment (gaoyang07) [reply mmrazor/engine/runner/evaluator_val_loop.py:98]: This method is called when it comes to those NAS algorithms that require building a supernet for training. For those algorithms, measuring subnet resources is more meaningful than supernet during validation, therefore this method is required to get the current searched subnet from the supernet.

File: mmrazor/engine/runner/evaluator_val_loop.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+from torch.utils.data import DataLoader
+
+from mmengine.evaluator import Evaluator
+from mmengine.runner import ValLoop
+from mmrazor.registry import ESTIMATORS, LOOPS
+from mmrazor.structures import ResourceEstimator
+
+
+@LOOPS.register_module()
+class EvaluatorLoop(ValLoop):
```
[the remark this answers is not in the document]
Comment (gaoyang07) [reply mmrazor/engine/runner/evaluator_val_loop.py]: Shall the file name be changed?

File: mmrazor/engine/runner/evaluator_val_loop.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        if hasattr(model, 'is_supernet') and model.is_supernet:
+            model = self.export_subnet(model)
+
+        resource_results = self.estimator.estimate(
+            model=model, resource_args=resource_args)
+
+        return resource_results
+
+    def export_subnet(self, model):
+        """Export current best subnet."""
```
[the remark this answers is not in the document]
Comment (gaoyang07) [reply mmrazor/engine/runner/evaluator_val_loop.py]: Done.

File: mmrazor/structures/estimator/resource_estimator.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        ...        module.__params__ += 700000
+        ...
+        >>> model = CustomModule()
+        >>> estimator.estimate(
+        ...     model=model,
+        ...     resource_args=dict(input_shape=(1, 3, 64, 64)))
+        {'flops': 1.0, 'params': 0.7, 'latency': 0.0}
+
+        >>> # calculate mmrazor.model flops
+        NOTE: check 'EvaluatorLoop' in engine.runner.evaluator_val_loop
```
[the remark this answers is not in the document]
Comment (gaoyang07) [reply mmrazor/structures/estimator/resource_estimator.py]: Done.

File: mmrazor/structures/estimator/op_spec_counters/base_counter.py (hunk 6 above)
Comment (gaoyang07) [reply mmrazor/structures/estimator/op_spec_counters/base_counter.py:5]: Done.

File: mmrazor/engine/runner/evaluator_val_loop.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sunnyxiaohu) [reply mmrazor/engine/runner/evaluator_val_loop.py]: yes

[the remark this answers is not in the document]
Comment (gaoyang07) [reply mmrazor/engine/runner/evaluator_val_loop.py]: Done.

File: mmrazor/engine/runner/resource_evaluator_val_loop.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+from mmengine.evaluator import Evaluator
+from mmengine.runner import ValLoop
+from torch.utils.data import DataLoader
+
+from mmrazor.registry import ESTIMATORS, LOOPS
+from mmrazor.structures import ResourceEstimator
+
+
+@LOOPS.register_module()
+class ResourceEvaluatorLoop(ValLoop):
```
[the remark this answers is not in the document]
Comment (humu789) [reply mmrazor/engine/runner/resource_evaluator_val_loop.py:14]: ResourceEvaluatorLoop seems to be replaced with Hook, thus we need not maintain source valloop.

**Review by humu789 — COMMENTED** [review]
1. It seems to lack the function of counting flops with the specified scope.
2. Better not use new registries without parents, it will be not used by other repos of OpenMMLab. You can use directly TASK_UTILS instead of `estimator` and `op_counter`
3. `estimator/`  is unsuitable to be under `structures/`.  Suggestion location : `mmrazor/models/task_modules/`.
4. The file structure of `estimator/` could be optimizer. Suggestion: 
a. add `couters/` in `estimator`
b. move `flops_params_counter.py` / `latency.py` / `op_spec_counters/` to `counters/`
c. rename `latency.py` to `latency_counter.py`
d. rename `estimator/` to `estimators/`

File: mmrazor/registry/registry.py [hunk 12]
```diff
@@ -111,3 +111,6 @@ def build_razor_model_from_cfg(
 VISUALIZERS = Registry('visualizer', parent=MMENGINE_VISUALIZERS)
 # manage visualizer backend
 VISBACKENDS = Registry('vis_backend', parent=MMENGINE_VISBACKENDS)
+
+ESTIMATORS = Registry('estimator')
+OP_SPEC_COUNTERS = Registry('op_counter')
```
[the remark this answers is not in the document]
Comment (gaoyang07) [reply mmrazor/registry/registry.py]: Done.

File: mmrazor/structures/estimator/resource_estimator.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        >>> estimator.estimate(
+        ...     model=model,
+        ...     resource_args=dict(
+        ...         input_shape=(1, 3, 64, 64),
+        ...         disabled_counters=['CustomModuleCounter']))
+        {'flops': 0.0, 'params': 0.0, 'latency': 0.0}
+
+        >>> # calculate resources of mmrazor.models
+        NOTE: check 'ResourceEvaluatorLoop' in 
+              engine.runner.resource_evaluator_val_loop for more details.
```
[the remark this answers is not in the document]
Comment (gaoyang07) [reply mmrazor/structures/estimator/resource_estimator.py]: Done.

File: mmrazor/engine/runner/resource_evaluator_val_loop.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (gaoyang07) [reply mmrazor/engine/runner/resource_evaluator_val_loop.py]: Make this part a hook, done.

**Comment by gaoyang07** [comment]
Now support counting flops with the specified scope. A list of scope names is required from users.

File: mmrazor/engine/runner/evolution_search_loop.py [hunk 14]
```diff
@@ -299,8 +301,15 @@ def _check_constraints(self, random_subnet: SupportRandomSubnet) -> bool:
 
         self.model.set_subnet(random_subnet)
         fix_mutable = export_fix_subnet(self.model)
-        flops: float = FlopsEstimator.get_model_complexity_info(
-            self.model, fix_mutable=fix_mutable, as_strings=False)[0]
+        copied_model = copy.deepcopy(self.model)
+        load_fix_subnet(copied_model, fix_mutable)
+
+        estimator = ResourceEstimator()
```
Comment (sunnyxiaohu) [inline mmrazor/engine/runner/evolution_search_loop.py:307]: Use estimator_cfg to build ResourceEstimator， and not use `input_shape` as fixed kwargs for `::estimate()`.

File: mmrazor/engine/runner/subnet_sampler_loop.py [hunk 15]
```diff
@@ -316,8 +318,15 @@ def _check_constraints(self, random_subnet: SupportRandomSubnet) -> bool:
 
         self.model.set_subnet(random_subnet)
         fix_mutable = export_fix_subnet(self.model)
-        flops = FlopsEstimator.get_model_complexity_info(
-            self.model, fix_mutable=fix_mutable, as_strings=False)[0]
+        copied_model = copy.deepcopy(self.model)
+        load_fix_subnet(copied_model, fix_mutable)
+
+        estimator = ResourceEstimator()
```
Comment (sunnyxiaohu) [inline mmrazor/engine/runner/subnet_sampler_loop.py:324]: Use estimator_cfg to build ResourceEstimator， and not use `input_shape` as fixed kwargs for `::estimate`.

File: mmrazor/models/task_modules/estimators/counters/flops_params_counter.py [hunk 16]
```diff
@@ -0,0 +1,467 @@
...
+
+import torch
+import torch.nn as nn
+
+from mmrazor.registry import TASK_UTILS
+
+
+def get_model_complexity_info(model,
+                              input_shape,
+                              spec_modules=[],
```
Comment (sunnyxiaohu) [inline mmrazor/models/task_modules/estimators/counters/flops_params_counter.py:13]: spec_modules -> custom_keys to support prefix, ref to mmcv::mmcv/runner/optimizer/default_constuctor.py

File: mmrazor/models/task_modules/estimators/counters/flops_params_counter.py [hunk 17]
```diff
@@ -0,0 +1,467 @@
...
+            flops_params_model,
+            flops_count,
+            params_count,
+            ost=ost,
+            flush=flush)
+
+    if len(spec_modules):
+        spec_modules_resources = dict()
+        accumulate_sub_module_flops_params(flops_params_model)
+        for name, module in flops_params_model.architecture.named_modules():
```
Comment (sunnyxiaohu) [inline mmrazor/models/task_modules/estimators/counters/flops_params_counter.py:102]: Not all the flops_params_model have the `architecture` attribute.

File: mmrazor/models/task_modules/estimators/counters/flops_params_counter.py [hunk 18]
```diff
@@ -0,0 +1,467 @@
...
+    def flops_repr(self):
+        accumulated_num_params = self.accumulate_params()
+        accumulated_flops_cost = self.accumulate_flops()
+        flops_string = str(
+            params_units_convert(
+                accumulated_flops_cost, units=units,
+                precision=precision)) + ' ' + units + 'FLOPs'
+        params_string = str(
+            params_units_convert(
+                accumulated_num_params, units='M',
```
Comment (sunnyxiaohu) [inline mmrazor/models/task_modules/estimators/counters/flops_params_counter.py:234]: Unify accumulated_flops_cost and accumulated_num_params with `units`

File: mmrazor/models/task_modules/estimators/counters/flops_params_counter.py [hunk 19]
```diff
@@ -0,0 +1,467 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import sys
+from functools import partial
+
+import torch
```
Comment (sunnyxiaohu) [inline mmrazor/models/task_modules/estimators/counters/flops_params_counter.py:5]: Unify `units` for params and flops under the scope of `flops_params_counter.py`

File: mmrazor/structures/subnet/estimators/flops.py [hunk 20]
```diff
@@ -93,12 +93,12 @@ class FlopsEstimator:
     def get_model_complexity_info(
             model: Module,
             fix_mutable: Optional[ValidFixMutable] = None,
-            input_shape: Iterable[int] = (3, 224, 224),
```
[the remark this answers is not in the document]
Comment (gaoyang07) [reply mmrazor/structures/subnet/estimators/flops.py:96]: Done.

File: mmrazor/models/task_modules/estimators/counters/flops_params_counter.py (hunk 18 above)
Comment (gaoyang07) [reply mmrazor/models/task_modules/estimators/counters/flops_params_counter.py:234]: A unit pair with FLOPs as 'G' and params as 'M' may be better.

File: mmrazor/engine/runner/evolution_search_loop.py (hunk 14 above)
Comment (gaoyang07) [reply mmrazor/engine/runner/evolution_search_loop.py]: Done.

File: mmrazor/engine/runner/subnet_sampler_loop.py (hunk 15 above)
Comment (gaoyang07) [reply mmrazor/engine/runner/subnet_sampler_loop.py]: Done.

File: mmrazor/models/task_modules/estimators/counters/flops_params_counter.py (hunk 17 above)
Comment (gaoyang07) [reply mmrazor/models/task_modules/estimators/counters/flops_params_counter.py]: Done.

File: mmrazor/models/task_modules/estimators/counters/flops_params_counter.py (hunk 16 above)
Comment (gaoyang07) [reply mmrazor/models/task_modules/estimators/counters/flops_params_counter.py:13]: Now support counting flops with a specified scope, e.g. spec_modules = ['backbone']

File: mmrazor/models/task_modules/estimators/counters/flops_params_counter.py (hunk 19 above)
Comment (gaoyang07) [reply mmrazor/models/task_modules/estimators/counters/flops_params_counter.py:5]: [cancel unit convert in accumulate_sub_module_flops_params](https://github.com/open-mmlab/mmrazor/pull/230/commits/335e9c4d17f91e09039b047f7b0d55cc9b22683c), remain the rest as the origin version.

**Pull request merged**